File: expense-agent/backend/app/models/expense.py
from datetime import datetime
from typing import TYPE_CHECKING, Optional

from sqlalchemy import DateTime, Float, ForeignKey, Integer, Numeric, String, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database.database import Base

if TYPE_CHECKING:
    from app.models.transaction import Transaction
    from app.models.user import User


class Expense(Base):
    __tablename__ = "expenses"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id"), nullable=False, default=1)
    transaction_id: Mapped[Optional[int]] = mapped_column(ForeignKey("transactions.id"), nullable=True)
    amount: Mapped[float] = mapped_column(Numeric(12, 2), nullable=False)
    merchant: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    category: Mapped[str] = mapped_column(String(80), nullable=False)
    subcategory: Mapped[Optional[str]] = mapped_column(String(80), nullable=True)
    description: Mapped[Optional[str]] = mapped_column(String(500), nullable=True)
    confidence: Mapped[Optional[float]] = mapped_column(Float, nullable=True)
    source: Mapped[str] = mapped_column(String(64), default="auto")
    direction: Mapped[str] = mapped_column(String(16), default="debit", index=True)  # debit | credit
    spent_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    user: Mapped["User"] = relationship(back_populates="expenses")
    transaction: Mapped[Optional["Transaction"]] = relationship(back_populates="expense")

File: expense-agent/frontend/types/expense.ts
export type Expense = {
  id: number;
  amount: number;
  merchant: string | null;
  category: string;
  subcategory: string | null;
  description: string | null;
  confidence: number | null;
  source: string;
  direction?: "debit" | "credit" | "transfer";
  spent_at: string;
  transaction_id: number | null;
  upi_ref?: string | null;
  reference_id?: string | null;
  payment_method?: string | null;
  card_issuer?: string | null;
  account_suffix?: string | null;
  created_at: string;
};
